test/unit/puppetlabs/services/jruby/jruby_pool_test.clj

```diff
@@ -99,6 +99,62 @@
               (doseq [k (keys counts)]
                 (is (= (inc (counts k)) (new-counts k)))))))))))
 
+(deftest borrow-while-pool-is-being-initialized-test
+  (testing "borrow will block until an instance is available while the pool is coming online"
+    (let [pool-initialized? (promise)
+          init-fn (fn [instance] @pool-initialized? instance)
+          pool-size 1
+          config (jruby-testutils/jruby-config
+                  {:max-active-instances pool-size
+                   :lifecycle {:initialize-pool-instance init-fn}})
+          pool-context (jruby-core/create-pool-context config)]
+
+      ;; start a pool initialization, which will block on the `initialize-pool-instance`
+      ;; function's deref of the promise
+      (jruby-agents/send-prime-pool! pool-context)
```

Comment:
It's probably worth a quick spike to see how hard it would be to port them to using `create-pool`.  If it doesn't seem too difficult then that sounds like a good idea to me.

I am also in favor of moving non-public functions out of `core` but we could ticket that as a separate task and not block this PR on it.